ubi_manifest/worker/tasks/depsolver/rpm_depsolver.py

```diff
@@ -0,0 +1,150 @@
+import os
+import logging
+from concurrent.futures import as_completed
+from pubtools.pulplib import Criteria
+from more_executors.futures import f_proxy
+from more_executors import Executors
+from .pulp_queries import search_modulemds, search_rpms
+from .utils import create_or_criteria, get_n_latest_from_content, parse_bool_deps
+
+_LOG = logging.getLogger(__name__)
+
+# need to set significantly lower batches for general rpm search
+# otherwise db may very likely hit OOM error.
+BATCH_SIZE_RPM = int(os.getenv("UBI_MANIFEST_BATCH_SIZE_RPM", "15"))
+BATCH_SIZE_RESOLVER = int(os.getenv("UBI_MANIFEST_BATCH_SIZE_RESOLVER", "150"))
+
+
+class Depsolver:
+    def __init__(self, repos):
+
+        self.repos = repos
+        self.output_set = set()
+
+        self._provides = set()  # set of all rpm.provides we've visited
+        self._requires = set()  # set of all rpm.requires we've visited
+
+        # set of solvables (pkg, lib, ...) that we use for checking remaining requires
+        self._unsolved = set()
+
+        self._modular_rpms = set()
+
+        self._executor = Executors.thread_pool(max_workers=4)
```

Comment:
If this class is going to create an executor, you should probably make Depsolver have `__enter__` / `__exit__` methods and make the `__exit__` method exit the executor, then use Depsolver via a `with` statement.

In cases where I didn't do that at first, I usually regretted it and had to go back & fix it later.  Even if it seems not to matter at first, it often ends up with some negative effects like: when the test suite runs, not all threads created by one test get shut down before starting the next, etc.